### PR TITLE
Satisfaction report

### PR DESCRIPTION
**Pull request by dwhenry**
Report for summarised satisfaction data. This report is a work in progress as
it still requires numbers for TPAS satisfaction and for Appointments completed 
for each delivery partner.

### CODE REVIEW

File: app/services/satisfaction_summary/average_column.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    end
+
+    private
+
+    def total_appointments
+      @total_appointments ||= @partners.sum(&:appointment_completions)
+    end
+
+    def weighted_average(field)
+      return 0 if total_appointments == 0
```
Comment (benlovell) [inline app/services/satisfaction_summary/average_column.rb]: `if total_appointments.zero?` mabes?

File: app/services/satisfaction_summary/average_column.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+      @partners.sum do |partner|
+        partner.send(field) * partner.appointment_completions
```
Comment (benlovell) [inline app/services/satisfaction_summary/average_column.rb]: Can we `public_send` here instead?

File: app/services/satisfaction_summary/partner_column.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    DELIGHTED = 4
+    VERY_PLEASED = 3
+    SATISFIED = 2
+    FRUSTRATED = 1
+    VERY_FRUSTRATED = 0
+
+    attr_reader :satisfaction_data
+
+    def initialize
+      @satisfaction_data = Hash.new(0)
```
Comment (benlovell) [inline app/services/satisfaction_summary/partner_column.rb]: The `0` argument is not necessary.

File: app/services/satisfaction_summary/partner_column.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        frustrated * FRUSTRATED +
+        very_frustrated * VERY_FRUSTRATED
+    end
+
+    def respondents
+      @satisfaction_data.values.sum
+    end
+
+    def gdx_average_score
+      return 0.0 unless respondents > 0
```
Comment (benlovell) [inline app/services/satisfaction_summary/partner_column.rb]: `respondents.positive?`

File: app/services/satisfaction_summary/partner_column.rb [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+      sum_of_score.to_f / (4 * respondents)
+    end
+
+    def top_two_score
+      return 0.0 unless respondents > 0
```
Comment (benlovell) [inline app/services/satisfaction_summary/partner_column.rb]: Same here.

File: app/services/satisfaction_summary/partner_column.rb [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    end
+
+    def respondents
+      @satisfaction_data.values.sum
+    end
+
+    def gdx_average_score
+      return 0.0 unless respondents > 0
+
+      sum_of_score.to_f / (4 * respondents)
```
Comment (benlovell) [inline app/services/satisfaction_summary/partner_column.rb]: What is the magic `4`?

File: app/services/satisfaction_summary/row.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        'weighted_average' => weighted_average
+      }
+    end
+
+    private
+
+    def value(column)
+      return unless column.respond_to?(@field)
+
+      column.send(@field)
```
Comment (benlovell) [inline app/services/satisfaction_summary/row.rb]: `public_send`?

File: app/services/satisfaction_summary/row.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    end
+
+    def format(value)
+      @formatter.send(@format, value)
```
Comment (benlovell) [inline app/services/satisfaction_summary/row.rb]: Same here possibly?

File: app/services/satisfaction_summary/total_column.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+      respondents
+      sum_of_score
+      appointment_completions
+    ).freeze
+
+    def initialize(partners)
+      @partners = partners
+    end
+
+    SUMMABLE_FIELDS.each do |field|
```
Comment (benlovell) [inline app/services/satisfaction_summary/total_column.rb]: Nice.

File: features/step_definitions/satisfaction_report_steps.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
+Given(/^there are existing satisfaction records$/) do
+  create_list(:satisfaction, 20)
```
Comment (benlovell) [inline features/step_definitions/satisfaction_report_steps.rb]: We could trim this down to just `11` I guess - and still be sure the other steps are testing what they need. Would be a little quicker.

File: features/satisfaction_report.feature [hunk 11]
```diff
@@ -0,0 +1,25 @@
+Feature: Satisfaction
+  As a Pension Wise data analyst
+  I want to know customer satisfaction
+  So that I can understand whether then servcie is performing to the required standard
```
Comment (benlovell) [inline features/satisfaction_report.feature]: Typo in 'service'

File: app/controllers/reports_controller.rb [hunk 12]
```diff
@@ -11,6 +11,30 @@ def call_volumes
...
   end
 
+  def satisfaction_summary
+    @satisfactions = Satisfactions.new(satisfaction_params)
+    @satisfaction_summary = SatisfactionSummary.new(@satisfactions.results)
+
+    respond_to do |format|
+      format.html
+      format.csv do
+        filename = "satisfaction_#{@satisfactions.period}"
```
Comment (benlovell) [inline app/controllers/reports_controller.rb]: Has there been an expressed requirement to include the period and stuff in the exported filenames? It adds a bit of noise and if it isn't particularly helpful to the data folks then we could simplify this a bit.

File: app/controllers/reports_controller.rb [hunk 13]
```diff
@@ -11,6 +11,30 @@ def call_volumes
...
+      end
+    end
+  end
+
+  def satisfaction
+    @satisfactions = Satisfactions.new(satisfaction_params)
+
+    respond_to do |format|
+      format.csv do
+        filename = "satisfaction_raw_#{@satisfactions.period}"
```
Comment (benlovell) [inline app/controllers/reports_controller.rb]: Same here, of course.

File: features/satisfaction_report.feature [hunk 14]
```diff
@@ -0,0 +1,25 @@
+Feature: Satisfaction
+  As a Pension Wise data analyst
```
Comment (benbarnett) [inline features/satisfaction_report.feature]: We've got [Pension Wise data analyst](https://github.com/guidance-guarantee-programme/reporting/blob/master/features/where_did_hear_about_us_report.feature#L2) and [data analyst](https://github.com/guidance-guarantee-programme/reporting/blob/master/features/call_volumes_report.feature#L2) across reporting here, are the same thing or intentionally different?

File: app/services/satisfaction_summary/formatter.rb [hunk 15]
```diff
@@ -0,0 +1,15 @@
+class SatisfactionSummary
+  class Formatter
+    def self.percent(value)
+      return '' if value.blank?
```
Comment (benbarnett) [inline app/services/satisfaction_summary/formatter.rb]: Would `0` be more useful here in terms of creating reports etc?

File: app/views/reports/satisfaction_summary.html.erb [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+        </tr>
+      </thead>
+      <tbody>
+        <%- @satisfaction_summary.rows.each do |row| %>
+          <tr class="t-row">
+            <td><%= row.name %></td>
+            <td class="text-right"><%= row.cas %></td>
+            <td class="text-right"><%= row.cita %></td>
+            <td class="text-right"><%= row.nicab %></td>
+            <td class="text-right"><%= row.tp %></td>
```
Comment (benbarnett) [inline app/views/reports/satisfaction_summary.html.erb]: Are we still planning to do the `tp` => `tpas` change separately? We're adding quite a lot of code that references it like this now.

File: app/services/satisfaction_summary/formatter.rb (hunk 15 above)
Comment (dwhenry) [inline app/services/satisfaction_summary/formatter.rb]: Come have a look at the report to see why I have done this and we can discuss.

File: app/views/reports/satisfaction_summary.html.erb (hunk 16 above)
Comment (dwhenry) [inline app/views/reports/satisfaction_summary.html.erb]: Good point, I meant to have the to change implemented before I completed this work, unfortunately when I spoke with John I needed to rethink delivery partners slightly and this has been left on hold.  Will change now.

File: features/satisfaction_report.feature (hunk 14 above)
Comment (dwhenry) [inline features/satisfaction_report.feature]: This was not intentional happy to standardise on one wording.

File: app/services/satisfaction_summary/average_column.rb (hunk 1 above)
Comment (dwhenry) [inline app/services/satisfaction_summary/average_column.rb]: done.

File: app/services/satisfaction_summary/partner_column.rb (hunk 3 above)
Comment (dwhenry) [inline app/services/satisfaction_summary/partner_column.rb]: as discussed this enabled a default value across the report

File: app/services/satisfaction_summary/partner_column.rb (hunk 4 above)
Comment (dwhenry) [inline app/services/satisfaction_summary/partner_column.rb]: used 

```
if respondents.zero?
```
instead

File: app/services/satisfaction_summary/partner_column.rb (hunk 6 above)
Comment (dwhenry) [inline app/services/satisfaction_summary/partner_column.rb]: good spot

File: features/step_definitions/satisfaction_report_steps.rb (hunk 10 above)
Comment (dwhenry) [inline features/step_definitions/satisfaction_report_steps.rb]: done - used 2 instead this is the minimum to test functionality

File: app/controllers/reports_controller.rb (hunk 12 above)
Comment (dwhenry) [inline app/controllers/reports_controller.rb]: done.

File: app/controllers/reports_controller.rb [hunk 17]
```diff
@@ -1,12 +1,33 @@
...
-    @call_volumes = CallVolumes.new(form_params)
+    @call_volumes = CallVolumes.new(date_params(:call_volumes))
 
     respond_to do |format|
       format.html
       format.csv do
-        filename = "call_volume_#{@call_volumes.period}"
-        render csv: DailyCallVolumeCsv.new(@call_volumes.results), filename: filename
+        render csv: DailyCallVolumeCsv.new(@call_volumes.results), filename: 'call_volume.csv'
+      end
```
Comment (dwhenry) [inline app/controllers/reports_controller.rb]: @benbarnett @benlovell Added a commit to standise CSV naming to include ".csv" and be the name of the report - after taking with John.

File: app/services/satisfaction_summary/row.rb [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+      @field = field
+      @format = format
+      @formatter = formatter
+    end
+
+    def name
+      @field.to_s.capitalize
+    end
+
+    def method_missing(method)
```
Comment (benlovell) [inline app/services/satisfaction_summary/row.rb]: You should implement `respond_to_missing?` here too.

Comment (dwhenry) [inline app/services/satisfaction_summary/row.rb]: replaced `repond_to?` with `respond_to_missing?`

**Pull request merged**